pkg/volumeclaimbinder/persistent_volume_claim_binder.go

```diff
@@ -0,0 +1,359 @@
+	binder.volumeIndex.Update(newVolume)
+	syncVolume(binder.volumeIndex, binder.client, newVolume)
+}
+
+func (binder *PersistentVolumeClaimBinder) deleteVolume(obj interface{}) {
+	binder.lock.Lock()
+	defer binder.lock.Unlock()
+	volume := obj.(*api.PersistentVolume)
+	binder.volumeIndex.Delete(volume)
+}
+
+func (binder *PersistentVolumeClaimBinder) addClaim(obj interface{}) {
+	binder.lock.Lock()
+	defer binder.lock.Unlock()
+	claim := obj.(*api.PersistentVolumeClaim)
+	syncClaim(binder.volumeIndex, binder.client, claim)
+}
+
+func (binder *PersistentVolumeClaimBinder) updateClaim(oldObj, newObj interface{}) {
+	binder.lock.Lock()
+	defer binder.lock.Unlock()
+	newClaim := newObj.(*api.PersistentVolumeClaim)
+	syncClaim(binder.volumeIndex, binder.client, newClaim)
+}
+
+func syncVolume(volumeIndex *persistentVolumeOrderedIndex, binderClient binderClient, volume *api.PersistentVolume) (err error) {
+	glog.V(5).Infof("Synchronizing PersistentVolume[%s]\n", volume.Name)
+
+	// volumes can be in one of the following states:
+	//
+	// VolumePending -- default value -- not bound to a claim and not yet processed through this controller.
+	// VolumeAvailable -- not bound to a claim, but processed at least once and found in this controller's volumeIndex.
+	// VolumeBound -- bound to a claim because volume.Spec.ClaimRef != nil.   Claim status may not be correct.
+	// VolumeReleased -- volume.Spec.ClaimRef != nil but the claim has been deleted by the user.
+	currentPhase := volume.Status.Phase
+	nextPhase := currentPhase
+
+	switch currentPhase {
+	// pending volumes are available only after indexing in order to be matched to claims.
+	case api.VolumePending:
+		_, exists, err := volumeIndex.Get(volume)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			volumeIndex.Add(volume)
+		}
+		glog.V(5).Infof("PersistentVolume[%s] is now available\n", volume.Name)
+		nextPhase = api.VolumeAvailable
+
+	// available volumes await a claim
+	case api.VolumeAvailable:
+		if volume.Spec.ClaimRef != nil {
+			_, err := binderClient.GetPersistentVolumeClaim(volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name)
+			if err == nil {
+				// change of phase will trigger an update event with the newly bound volume
+				glog.V(5).Infof("PersistentVolume[%s] is now bound\n", volume.Name)
+				nextPhase = api.VolumeBound
+			} else {
+				if errors.IsNotFound(err) {
+					nextPhase = api.VolumeReleased
+				}
+			}
+		}
+	//bound volumes require verification of their bound claims
+	case api.VolumeBound:
+		if volume.Spec.ClaimRef == nil {
+			return fmt.Errorf("PersistentVolume[%s] expected to be bound but found nil claimRef: %+v", volume)
+		} else {
+			claim, err := binderClient.GetPersistentVolumeClaim(volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name)
+			if err == nil {
+				// bound and active.  Build claim status as needed.
+				if claim.Status.VolumeRef == nil {
+					syncClaimStatus(binderClient, volume, claim)
+				}
```

Comment:
Do you mean that I should have code that validates that the volume just bound to is the volume passed into the func?  Or do you mean that the claim should check that its bound volume still exists?

If the latter, that would go in the control loop for the claim.  I had a check like that very early on but thought it would be an egregious violation of protocol to have a volume deleted from beneath a claim and that the error would come up through the pod.